src/VisualStudio/Core/Def/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -192,160 +179,152 @@ private WatcherOperation(Kind kind, List<uint> cookies)
-                        return true;
-
-                    case Kind.UnwatchFiles when other._kind == Kind.UnwatchFiles:
-                        combined = UnwatchFiles(_tokens.ToImmutableList().AddRange(other._tokens));
-                        return true;
+                for (; start <= end; start++)
+                {
+                    var op = operations[start];
 
-                    default:
-                        break;
+                    switch (op._kind)
+                    {
+                        case Kind.WatchFiles:
+                            for (var i = 0; i < op._paths.Count; i++)
+                            {
+                                fileNamesBuilder.AddRange(op._paths[i]);
+                            }
+
+                            for (var i = 0; i < op._tokens.Count; i++)
+                            {
+                                tokensBuilder.AddRange(op._tokens[i]);
+                            }
+                            break;
+
+                        case Kind.UnwatchFiles:
+                            for (var i = 0; i < op._tokens.Count; i++)
+                            {
+                                tokensBuilder.AddRange(op._tokens[i]);
+                            }
+                            break;
+
+                        case Kind.UnwatchDirectories:
+                            cookiesBuilder.AddRange(op._cookies);
+                            break;
+
+                        default:
+                            break;
+                    }
                 }
 
-                combined = default;
-                return false;
+                return firstOp._kind switch
+                {
+                    Kind.WatchFiles =>
+                        WatchFiles(fileNamesBuilder.ToImmutable(), firstOp._fileChangeFlags, firstOp._sink, tokensBuilder.ToImmutable()),
+                    Kind.UnwatchFiles =>
+                        UnwatchFiles(tokensBuilder.ToImmutable()),
+                    Kind.UnwatchDirectories =>
+                        UnwatchDirectories(cookiesBuilder.ToList()),
+                    _ =>
+                        throw ExceptionUtilities.Unreachable()
+                };
             }
 
-            public async ValueTask ApplyAsync(IVsAsyncFileChangeEx service, CancellationToken cancellationToken)
+            public bool CanCombineWith(in WatcherOperation other)
             {
-                switch (_kind)
+                return _kind switch
                 {
-                    case Kind.None:
-                        return;
+                    Kind.WatchDirectory => false, // Watching directory operation cannot be combined
+                    Kind.WatchFiles when other._kind is Kind.WatchFiles => true,
+                    Kind.UnwatchFiles when other._kind is Kind.UnwatchFiles => true,
+                    Kind.UnwatchDirectories when other._kind == Kind.UnwatchDirectories => true,
```

Comment:
At this point this is now just combinable if the kinds are equal, right? (and a special case for WatchDirectory)

Reply:
Sure, I can make that change.